### PR TITLE
[TIMOB-24174]: Android: Update ListView check/disclosure/child graphics

### PR DESCRIPTION
**Pull request by m1ga**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-24174

* update the old PR with ListView graphics
* bigger margin on left side of the labels
* bigger default height in a default ListView

### CODE REVIEW

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java [hunk 1]
```diff
@@ -439,6 +441,36 @@ public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCoun
 		setNativeView(wrapper);
 	}
 	
+	private int getImageRessource(String img){
```
Comment (garymathews) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java]: This should be moved into `TiRHelper` so we can use this elsewhere 👍

**Comment by hansemannn** [comment]
bildschirmfoto 2017-02-13 um 14 14 02

They should retain the old color, otherwise developers might complain about the change.

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java [hunk 2]
```diff
@@ -439,6 +441,36 @@ public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCoun
...
+		//
+		String resName = "drawable."+img+"_48"; // default medium
```
Comment (hansemannn) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java]: Spaces: ` + img + `

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java [hunk 3]
```diff
@@ -43,7 +43,7 @@
 	// every time we add a row. No sense checking it each time.
 	private static boolean ICS_OR_GREATER = (Build.VERSION.SDK_INT >= TiC.API_LEVEL_ICE_CREAM_SANDWICH);
 
-	private static final int LEFT_MARGIN = 5;
+	private static final int LEFT_MARGIN = 22;
```
Comment (hansemannn) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java]: Does Android natively has 22dp left and 27dp right?

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java (hunk 1 above)
Comment (m1ga) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java]: Moved it to TiRHelper

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java (hunk 3 above)
Comment (m1ga) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java]: A bit hard to find out as I find various examples (from 0 margin to 48). But I found this theme.xml:
https://github.com/android/platform_frameworks_base/blob/82834baa358f55acb542e17da828b2d497cf8332/core/res/res/values/themes.xml#L146
and it looks like 6dp is a default value for listitems in a basic theme. I will update ListView and TableView to look the same

Comment (hansemannn) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java]: ✅

Comment (m1ga) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java]: finally solved it :) Stupid dp vs pixels. Now both images (listview/tableview) have a 6dp padding

**Comment by hansemannn** [comment]
@garymathews or @fmerzadyan Can one of you review this? Thxxxx!

**Review by garymathews — APPROVED** [review]
CR: **PASS**
FT: **PASS**

###### TEST CASE
```JS
var win = Ti.UI.createWindow({
        backgroundColor: 'gray',
        layout: 'vertical'
    }),
    tv = Ti.UI.createTableView({
        headerTitle: 'TableView',
        data: [
            {
                title: 'Apples',
                hasChild: true
            }, {
                title: 'Bananas'
            }
        ],
        height: Ti.UI.SIZE
    }),
    ls = Ti.UI.createListSection({
        headerTitle: 'ListView'
    }),
    lv = Ti.UI.createListView({
        height: Ti.UI.SIZE
    });

ls.setItems([
    {
        properties: {
            title: 'Apples',
            accessoryType: Ti.UI.LIST_ACCESSORY_TYPE_CHECKMARK
        }
    }, {
        properties: {
           title: 'Bananas',
           accessoryType: Ti.UI.LIST_ACCESSORY_TYPE_DETAIL
       }
    }
]);
lv.sections = [ls];

win.add(tv);
win.add(lv);
win.open();
```

| BEFORE | AFTER |
| --- | --- |
|  |  |

👍

**Comment by lokeshchdhry** [comment]
@hansemannn , Can you please rebase/update this branch, so that merge is enabled. Thanks.

**Comment by hansemannn** [comment]
@lokeshchdhry done!

**Comment by hansemannn** [comment]
@m1ga We need a backport for this asap to include it in 6.1.0. Should I do or can you?

**Comment by lokeshchdhry** [comment]
@hansemannn ,Cool thanks. I still see the merge not enabled due to statuses/checks did not pass.

**Comment by lokeshchdhry** [comment]
@m1ga, @hansemannn, @garymathews - I am seeing some differences in phone & emulator between tableview & listview icon sizes.

One Plus one - 6.0.1 | Emulator android 7.1.1
------------ | -------------
android_6 0 1-one_plus_one |  emu_android_7 1 1

If you see the apple icon in table view is slightly larger than the 2 list view icons where as in the emulator the tableview icon is smaller compared to the listview icons.

**Comment by m1ga** [comment]
@lokeshchdhry No, that shouldn't be. I'll check that again, thanks for the feedback!

**Comment by m1ga** [comment]
No luck in finding the culprit at the moment. Guess it's somewhere in `TiTableViewRowProxyItem.java` around line 552. So I need more time to test it

And I have a strange crash on 5.0.2, that `/android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListItem.java` #62 s null:

`04-19 21:13:00.387: E/AndroidRuntime(12800): java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.ImageView.setImageResource(int)' on a null object reference
04-19 21:13:00.387: E/AndroidRuntime(12800): 	at ti.modules.titanium.ui.widget.listview.TiListItem.handleAccessory(TiListItem.java:62)
`

**Comment by m1ga** [comment]
oh no :-( that was a wrong commit

**Comment by m1ga** [comment]
Sorry about that. Should be fine again.

@lokeshchdhry the size of the tableview image was different (not 17dp like the listview) and one image was missing (disclosure_96.png). That is fixed now and tested on a Nexus 6 (emulator), Moto G and HTC A9.

**Comment by hansemannn** [comment]
Thanks Michael! I'm wondering why the header title background-color is different between tableview and listview?

**Comment by m1ga** [comment]
Which one should I make the default one? Dark or light one?

TableView: https://github.com/appcelerator/titanium_mobile/blob/a8aceeaeee02ed9c8ea70c835a0e78960e254a55/android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewHeaderItem.java#L54

ListView: https://github.com/appcelerator/titanium_mobile/blob/a8aceeaeee02ed9c8ea70c835a0e78960e254a55/android/modules/ui/res/layout/titanium_ui_list_header_or_footer.xml#L7

**Comment by hansemannn** [comment]
What ever is used natively :-) All sizes, background-colors, text colors etc pp. Those values are from old times. Maybe we do those in 7.0.0, but better soon than later.

**Comment by m1ga** [comment]
There is no real native version of this. In a ListView you would just create a View and add it (and style it yourself), so like the Ti-ListView is doing already. And you can already use custom views so the headerTitle is just a short cut.

The closest thing I could find in the Material style guide is:

But that should be a different (bigger) PR because you have to adjust the default Android theme (e.g. not use the dark one by default).

For this PR I'll use the bright version in both headerTitles so it looks fresher and a bit more like the iOS version.

**Comment by hansemannn** [comment]
Yeah, let's keep that part as it is. Thanks @m1ga!

**Comment by m1ga** [comment]
Headers look the same now:

device-2017-04-23-124704

**Review by garymathews — APPROVED** [review]
CR: **PASS**
FT: **PASS**

**Comment by lokeshchdhry** [comment]
@garymathews , @m1ga , I am getting this crash when I run test code above with locally built 6.2.0 on android API 24 7.0.0 & above.

```
[ERROR] :  TiApplication: (main) [732,732] Sending event: exception on thread: main msg:java.lang.ClassCastException: ti.modules.titanium.ui.TableViewSectionProxy cannot be cast to ti.modules.titanium.ui.TableViewRowProxy; Titanium 6.2.0,2017/06/13 09:56,undefined
[ERROR] :  TiApplication: java.lang.ClassCastException: ti.modules.titanium.ui.TableViewSectionProxy cannot be cast to ti.modules.titanium.ui.TableViewRowProxy
[ERROR] :  TiApplication: 	at ti.modules.titanium.ui.widget.tableview.TiTableView$TTVListAdapter.getView(TiTableView.java:205)
[ERROR] :  TiApplication: 	at android.widget.AbsListView.obtainView(AbsListView.java:2360)
[ERROR] :  TiApplication: 	at android.widget.ListView.measureHeightOfChildren(ListView.java:1326)
[ERROR] :  TiApplication: 	at android.widget.ListView.onMeasure(ListView.java:1233)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6120)
[ERROR] :  TiApplication: 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at org.appcelerator.titanium.view.TiCompositeLayout.constrainChild(TiCompositeLayout.java:445)
[ERROR] :  TiApplication: 	at org.appcelerator.titanium.view.TiCompositeLayout.onMeasure(TiCompositeLayout.java:303)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at org.appcelerator.titanium.view.TiCompositeLayout.constrainChild(TiCompositeLayout.java:445)
[ERROR] :  TiApplication: 	at org.appcelerator.titanium.view.TiCompositeLayout.onMeasure(TiCompositeLayout.java:303)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6120)
[ERROR] :  TiApplication: 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
[ERROR] :  TiApplication: 	at android.support.v7.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:135)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6120)
[ERROR] :  TiApplication: 	at android.support.v7.widget.ActionBarOverlayLayout.onMeasure(ActionBarOverlayLayout.java:391)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6120)
[ERROR] :  TiApplication: 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at android.widget.LinearLayout.measureVertical(LinearLayout.java:911)
[ERROR] :  TiApplication: 	at android.widget.LinearLayout.onMeasure(LinearLayout.java:640)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6120)
[ERROR] :  TiApplication: 	at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
[ERROR] :  TiApplication: 	at com.android.internal.policy.DecorView.onMeasure(DecorView.java:687)
[ERROR] :  TiApplication: 	at android.view.View.measure(View.java:19730)
[ERROR] :  TiApplication: 	at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:2271)
[ERROR] :  TiApplication: 	at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:1358)
[ERROR] :  TiApplication: 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1607)
[ERROR] :  TiApplication: 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1246)
[ERROR] :  TiApplication: 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6301)
[ERROR] :  TiApplication: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:871)
[ERROR] :  TiApplication: 	at android.view.Choreographer.doCallbacks(Choreographer.java:683)
[ERROR] :  TiApplication: 	at android.view.Choreographer.doFrame(Choreographer.java:619)
[ERROR] :  TiApplication: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:857)
[ERROR] :  TiApplication: 	at android.os.Handler.handleCallback(Handler.java:751)
[ERROR] :  TiApplication: 	at android.os.Handler.dispatchMessage(Handler.java:95)
[ERROR] :  TiApplication: 	at android.os.Looper.loop(Looper.java:154)
[ERROR] :  TiApplication: 	at android.app.ActivityThread.main(ActivityThread.java:6077)
[ERROR] :  TiApplication: 	at java.lang.reflect.Method.invoke(Native Method)
[ERROR] :  TiApplication: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
[ERROR] :  TiApplication: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

**Comment by hansemannn** [comment]
@lokeshchdhry Does it crash with master (non-PR) as well? Wondering because there is no proxy-change related to the crash.

**Comment by m1ga** [comment]
I'll have a look at that again! It might be related to this PR: https://github.com/appcelerator/titanium_mobile/pull/8948 ?

**Comment by lokeshchdhry** [comment]
@hansemannn , Yes I get the same error with a non-PR build as well when I run the test code above.

**Comment by hansemannn** [comment]
@garymathews might be the Android 7 PR then.

**Comment by hansemannn** [comment]
@lokeshchdhry Can you revalidate that now that the list-view issue is fixed?

**Comment by garymathews** [comment]
@hansemannn @lokeshchdhry https://github.com/appcelerator/titanium_mobile/pull/9144 needs to be merged before this can be reviewed (to fix the crash bug).

**Comment by jquick-axway** [comment]
@m1ga, I know I'm late to the party here, but I'd like to add some feedback.

I don't think we should be custom loading these drawable images via the _48, _72, _96, etc. postfix notation. Instead, it would be better to take advantage of Android's built-in density aware drawable folders (ex: drawable-mdpi, drawable-hdpi, etc.) since the Android OS will automatically load the correct density scaled drawable for us... or the closest match in case we're running on under a DPI that we don't have an image. This will help make it more future proof for DPIs that we current do not support, such as tvdpi (sometimes used by tablets like the 1st generation Nexus 7), and Android will also auto-scale the drawable up to match the current DPI. An additional advantage of this approach is that you can then always assume a drawable will always be loaded as "48dp" (the mdpi base resolution), even if you don't have a matching drawable matching the device's DPI.

So, I recommend we do the following:

- Move the UI module's drawable images to the following **new** directories:
  * `drawable/*_36.png` -> `drawable-ldpi`
  * `drawable/*_48.png` -> `drawable-mdpi`
  * `drawable/*_72.png` -> `drawable-hdpi`
  * `drawable/*_96.png` -> `drawable-xhdpi`
  * `drawable/*_144.png` -> `drawable-xxhdpi`
  * `drawable/*_192.png` -> `drawable-xxxhdpi`
- Remove all the postfixed `_###` parts from the file names.
- Delete the UI module's `res/drawable` directory.
- Remove the TiRHelper.getImageRessource() method you've added.
  * The way it's loading drawables is non-standard. Might confuse developers.
- Acquire these drawable images' resource IDs as follows:
  * `int resId = TiRHelper.getResource("drawable." + imageName, false);`

Sorry to picky about this. I know you didn't come up with this custom image loading scheme. But I think this is the best way to resolve this for good. And I appreciate you take a stab at this.

**Comment by jquick-axway** [comment]
Also, if we do what I suggested above, we may want to prefix these drawable images with "titanium_*.png" to avoid name collision with the Titanium developer's own resource image files. Standard practice on Android is to use a C-style prefixing notation like this. (Ex: Facebook SDK uses a "fb_" prefix for its resources.)

**Comment by m1ga** [comment]
@jquick-axway Sounds like a good idea! I'll have a look at this, thank you for the feedback

**Pull request merged**